File: src/Evalution/AstBuilder.fs
module Evalution.AstBuilder
    open Piglet.Parser

    let build (input:string) : Ast.Program =
        let configurator = ParserFactory.Configure<obj>()

        let expressionSpec = configurator.CreateNonTerminal()
        let multiCallExpressionSpec = configurator.CreateNonTerminal()
        let unaryExpressionSpec = configurator.CreateNonTerminal()
        let argumentsSpec = configurator.CreateNonTerminal()
        let optionalArgumentsSpec = configurator.CreateNonTerminal()

        let int32Literal = configurator.CreateTerminal(@"\d+", fun x -> box (Ast.Int32Literal (int x) ) )
        let doubleLiteral = configurator.CreateTerminal(@"\d+\.\d+", fun x -> box (Ast.DoubleLiteral (float x) ) )
        let plus = configurator.CreateTerminal(@"\+")
        let minus = configurator.CreateTerminal(@"-")
        let exclamation = configurator.CreateTerminal(@"!")
        let asterisk = configurator.CreateTerminal(@"\*")
        let forwardSlash = configurator.CreateTerminal(@"/")
        let dot = configurator.CreateTerminal(@"\.")
        let comma = configurator.CreateTerminal(@",")
        let openSquare = configurator.CreateTerminal(@"\[")
        let closeSquare = configurator.CreateTerminal(@"\]")
        let openParen = configurator.CreateTerminal(@"\(")
        let closeParen = configurator.CreateTerminal(@"\)")
        let identifier = configurator.CreateTerminal(@"[a-zA-Z_][a-zA-Z_0-9]*", fun x -> box(x))
        let timeSpan = configurator.CreateTerminal(@"TimeSpan\.FromHours")

        configurator.LeftAssociative(plus, minus, exclamation) |> ignore
        configurator.LeftAssociative(asterisk) |> ignore
        configurator.LeftAssociative(forwardSlash) |> ignore
        configurator.LeftAssociative(dot) |> ignore
        let unaryExpressionPrecedenceGroup  = configurator.RightAssociative()

        // Parens
        expressionSpec.AddProduction(openParen, expressionSpec, closeParen)
            .SetReduceFunction(fun x -> box (x.[1] :?> Ast.Expression))


        // BinaryExpressions
        expressionSpec.AddProduction(expressionSpec, plus, expressionSpec)
            .SetReduceFunction(fun x -> box (Ast.BinaryExpression(x.[0] :?> Ast.Expression, Ast.Add, x.[2] :?>Ast.Expression)))
        expressionSpec.AddProduction(expressionSpec, minus, expressionSpec)
            .SetReduceFunction(fun x -> box (Ast.BinaryExpression(x.[0] :?> Ast.Expression, Ast.Subtract, x.[2] :?>Ast.Expression)))
        expressionSpec.AddProduction(expressionSpec, asterisk, expressionSpec)
            .SetReduceFunction(fun x -> box (Ast.BinaryExpression(x.[0] :?> Ast.Expression, Ast.Multiply, x.[2] :?>Ast.Expression)))
        expressionSpec.AddProduction(expressionSpec, forwardSlash, expressionSpec)
            .SetReduceFunction(fun x -> box (Ast.BinaryExpression(x.[0] :?> Ast.Expression, Ast.Divide, x.[2] :?>Ast.Expression)))

        // Literals
        expressionSpec.AddProduction(int32Literal)
            .SetReduceFunction(fun x -> box (Ast.LiteralExpression(x.[0]:?>Ast.Literal ) ))
        expressionSpec.AddProduction(doubleLiteral)
            .SetReduceFunction(fun x -> box (Ast.LiteralExpression(x.[0]:?>Ast.Literal ) ))

        // UnaryOperators
        let expressionProduction = expressionSpec.AddProduction(unaryExpressionSpec, expressionSpec)
        expressionProduction.SetReduceFunction(fun x -> box(Ast.UnaryExpression(x.[0] :?>Ast.UnaryOperator, x.[1] :?> Ast.Expression)))
        expressionProduction.SetPrecedence unaryExpressionPrecedenceGroup

        unaryExpressionSpec.AddProduction(exclamation)
            .SetReduceFunction(fun x -> box(Ast.LogicalNegate))
        unaryExpressionSpec.AddProduction(minus)
            .SetReduceFunction(fun x -> box(Ast.Negate))
        unaryExpressionSpec.AddProduction(plus)
            .SetReduceFunction(fun x -> box(Ast.Identity))


        // Multicall expression
        expressionSpec.AddProduction(multiCallExpressionSpec)
            .SetReduceFunction(fun x -> box (Ast.MultiCallExpression(x.[0] :?> Ast.Multicall)) )

        // Default context method call
        multiCallExpressionSpec.AddProduction(identifier, openParen, optionalArgumentsSpec, closeParen)
            .SetReduceFunction(fun x -> box (Ast.CurrentContextMethodCall(
                                                x.[0] :?> string,   
                                                x.[2] :?> Ast.Arguments)) )
        // Object context method call
        multiCallExpressionSpec.AddProduction(multiCallExpressionSpec, dot, identifier, openParen, optionalArgumentsSpec, closeParen)
            .SetReduceFunction(fun x -> box (Ast.ObjectContextMethodCall(
                                                x.[0] :?>Ast.Multicall,
                                                x.[2] :?> string,
                                                x.[4] :?> Ast.Arguments)) )

        // Property call
        multiCallExpressionSpec.AddProduction(identifier)
            .SetReduceFunction(fun x -> box (Ast.CurrentContextPropertyCall(x.[0] :?> string)) )

        // Subproperty call
        multiCallExpressionSpec.AddProduction(multiCallExpressionSpec, dot, identifier)
            .SetReduceFunction(fun x -> box (Ast.ObjectContextPropertyCall(x.[0] :?>Ast.Multicall, x.[2] :?> string)) )

        // Array element call
        multiCallExpressionSpec.AddProduction(multiCallExpressionSpec, openSquare, expressionSpec, closeSquare)
            .SetReduceFunction(fun x -> box (Ast.ArrayElementCall(
                                                x.[0] :?> Ast.Multicall,
                                                x.[2] :?> Ast.Expression
                                                )) )
        
        optionalArgumentsSpec.AddProduction(argumentsSpec).SetReduceToFirst()
        optionalArgumentsSpec.AddProduction()
            .SetReduceFunction(fun x -> box (List.empty<Ast.Expression>))

        argumentsSpec.AddProduction(argumentsSpec, comma, expressionSpec)
            .SetReduceFunction(fun x -> box(
                        (x.[0] :?> Ast.Expression list)
                        @
                        ([x.[2] :?> Ast.Expression])
                )
            )
        argumentsSpec.AddProduction(expressionSpec)
            .SetReduceFunction(fun x -> box ([x.[0] :?> Ast.Expression]))

        
        let parser = configurator.CreateParser()
        let result = parser.Parse(input)

        result :?> Ast.Program
        

File: src/EvalutionCS/Ast/AstBuilder.cs
using System;
using System.Collections.Generic;

namespace Evalution.Ast
{
    using Piglet.Parser;

    public class AstBuilder
    {
        public static Expression Build(string input)
        {
            var configurator = ParserFactory.Configure<object>();

            var expressionSpec = configurator.CreateNonTerminal();
            var multiCallExpressionSpec = configurator.CreateNonTerminal();
            var unaryExpressionSpec = configurator.CreateNonTerminal();
            var argumentsSpec = configurator.CreateNonTerminal();
            var optionalArgumentsSpec = configurator.CreateNonTerminal();

            var int32Literal = configurator.CreateTerminal(@"\d+", x => new Int32Literal(Int32.Parse(x)));
            var doubleLiteral = configurator.CreateTerminal(@"\d+\.\d+", x => new DoubleLiteral(Double.Parse(x)) );
            var plus = configurator.CreateTerminal(@"\+");
            var minus = configurator.CreateTerminal(@"-");
            var exclamation = configurator.CreateTerminal(@"!");
            var asterisk = configurator.CreateTerminal(@"\*");
            var forwardSlash = configurator.CreateTerminal(@"/");
            var dot = configurator.CreateTerminal(@"\.");
            var comma = configurator.CreateTerminal(@",");
            var openSquare = configurator.CreateTerminal(@"\[");
            var closeSquare = configurator.CreateTerminal(@"\]");
            var openParen = configurator.CreateTerminal(@"\(");
            var closeParen = configurator.CreateTerminal(@"\)");
            var identifier = configurator.CreateTerminal(@"[a-zA-Z_][a-zA-Z_0-9]*", x => x);
            var timeSpan = configurator.CreateTerminal(@"TimeSpan\.FromHours");

            configurator.LeftAssociative(plus, minus, exclamation);
            configurator.LeftAssociative(asterisk);
            configurator.LeftAssociative(forwardSlash);
            configurator.LeftAssociative(dot);
            var unaryExpressionPrecedenceGroup = configurator.RightAssociative();

            // Parens
            expressionSpec.AddProduction(openParen, expressionSpec, closeParen)
                .SetReduceFunction(x => (Expression)x[1]);

            // BinaryExpressions
            expressionSpec.AddProduction(expressionSpec, plus, expressionSpec)
                .SetReduceFunction(x => new BinaryExpression((Expression)x[0], BinaryOperator.Add, (Expression)x[2]));
            expressionSpec.AddProduction(expressionSpec, minus, expressionSpec)
                .SetReduceFunction(x => new BinaryExpression((Expression)x[0], BinaryOperator.Subtract, (Expression)x[2]));
            expressionSpec.AddProduction(expressionSpec, asterisk, expressionSpec)
                .SetReduceFunction(x => new BinaryExpression((Expression)x[0], BinaryOperator.Multiply, (Expression)x[2]));
            expressionSpec.AddProduction(expressionSpec, forwardSlash, expressionSpec)
                .SetReduceFunction(x => new BinaryExpression((Expression)x[0], BinaryOperator.Divide, (Expression)x[2]));

            // Literals
            expressionSpec.AddProduction(int32Literal)
                .SetReduceFunction(x => new LiteralExpression((Literal)x[0]) );
            expressionSpec.AddProduction(doubleLiteral)
                .SetReduceFunction(x => new LiteralExpression((Literal)x[0]) );

            // UnaryOperators
            var expressionProduction = expressionSpec.AddProduction(unaryExpressionSpec, expressionSpec);
            expressionProduction.SetReduceFunction(x => new UnaryExpression((UnaryOperator)x[0], (Expression)x[1]));
            expressionProduction.SetPrecedence(unaryExpressionPrecedenceGroup);

            unaryExpressionSpec.AddProduction(exclamation)
                .SetReduceFunction(x => UnaryOperator.LogicalNegate);
            unaryExpressionSpec.AddProduction(minus)
                .SetReduceFunction(x => UnaryOperator.Negate);
            unaryExpressionSpec.AddProduction(plus)
                .SetReduceFunction(x => UnaryOperator.Identity);

            // Multicall expression
            expressionSpec.AddProduction(multiCallExpressionSpec)
                .SetReduceFunction(x => new MultiCallExpression((Multicall) x[0]));

             // Default context method call
            multiCallExpressionSpec.AddProduction(identifier, openParen, optionalArgumentsSpec, closeParen)
                .SetReduceFunction(x => new CurrentContextMethodCall(
                    (string) x[0],
                    (List<Expression>) x[2]));

            // Object context method call
            multiCallExpressionSpec.AddProduction(multiCallExpressionSpec, dot, identifier, openParen, optionalArgumentsSpec, closeParen)
                .SetReduceFunction(x => new ObjectContextMethodCall(
                                                    (Multicall)x[0],
                                                    (string)x[2],
                                                    (List<Expression>)x[4]));

            // Property call
            multiCallExpressionSpec.AddProduction(identifier)
                .SetReduceFunction(x => new CurrentContextPropertyCall((string) x[0]));

            // Subproperty call
            multiCallExpressionSpec.AddProduction(multiCallExpressionSpec, dot, identifier)
                .SetReduceFunction(x => new ObjectContextPropertyCall((Multicall) x[0], (string) x[2]));

            // Array element call
            multiCallExpressionSpec.AddProduction(multiCallExpressionSpec, openSquare, expressionSpec, closeSquare)
                .SetReduceFunction(x => new ArrayElementCall(
                                                    (Multicall)x[0],
                                                    (Expression)x[2]));

            optionalArgumentsSpec.AddProduction(argumentsSpec).SetReduceToFirst();
            optionalArgumentsSpec.AddProduction()
                .SetReduceFunction(x =>new List<Expression>());

            argumentsSpec.AddProduction(argumentsSpec, comma, expressionSpec)
                .SetReduceFunction(x => AddElementToList((List<Expression>)x[0], (Expression)x[2]));

            argumentsSpec.AddProduction(expressionSpec)
                .SetReduceFunction(x => new List<Expression>() { (Expression)x[0] });

            var parser = configurator.CreateParser();
            var result = parser.Parse(input);

            return (Expression)result;
        }

        private static List<Expression> AddElementToList(List<Expression> expressions, Expression expression)
        {
            expressions.Add(expression);
            return expressions;
        }
    }
}
